scripts/check_compliance.py

```diff
@@ -426,73 +426,123 @@ class Codeowners(ComplianceTest):
-                        if not g:
-                            self.add_failure("Path '{}' not found, in "
-                                             "CODEOWNERS".format(path))
-                        else:
-                            files = []
-                            if not add_base:
-                                for f in g:
-                                    l = f.replace(git_root + "/", "")
-                                    files.append(l)
-                            else:
-                                files = g
-
-                            all_files += files
-
-        files = []
-        for f in all_files:
-            if os.path.isfile(f):
-                files.append(f)
-
-        return set(files)
+            linenum = 0
+            for line in codeo:
+                linenum += 1
+                if line.startswith("#") or not line.strip():
+                    continue
+
+                match = re.match(r"^([^\s]+)\s+[^\s]+", line)
+                if not match:
+                    self.add_failure(
+                        "Invalid CODEOWNERS line %d\n\t%s" %
+                        (linenum, line))
+                    continue
+
+                git_patrn = match.group(1)
+                glob = self.git_pattern_to_glob(git_patrn)
+                files = []
+                for p in top_path.glob(glob):
+                    # comparing strings is much faster later
+                    files.append(str(p)[top_len:])
+
+                if not files:
+                    self.add_failure("Path '{}' not found, in "
+                                     "CODEOWNERS".format(git_patrn))
+
+                pattern2files[git_patrn] = files
+
+        return pattern2files
+
+
+    def git_pattern_to_glob(self, git_pattern):
+        """Appends and prepends '**[/*]' when needed. Result has neither a
+        leading nor a trailing slash.
+        """
+
+        ret = git_pattern
+
+        if git_pattern.startswith("/"):
+            ret = ret[1:]
+        else:
+            ret = "**/" + ret
+
+        if git_pattern.endswith("/"):
+            ret = ret + "**/*"
+        elif os.path.isdir(os.path.join(git_top, ret)):
+            self.add_failure("Expected '/' after directory '{}' "
+                             "in CODEOWNERS".format(ret))
+
+        return ret
+
 
     def run(self):
         self.prepare()
-        git_root = sh.git("rev-parse", "--show-toplevel").strip()
-        codeowners = os.path.join(git_root, "CODEOWNERS")
+        # TODO: testing an old self.commit range that doesn't end
+        # with HEAD is most likely a mistake. Should warn, see
+        # https://github.com/zephyrproject-rtos/ci-tools/pull/24
+        codeowners = os.path.join(git_top, "CODEOWNERS")
         if not os.path.exists(codeowners):
             self.skip("CODEOWNERS not available in this repo")
 
         commit = sh.git("diff","--name-only", "--diff-filter=A", self.commit_range, **sh_special_args)
```

Comment:
Missing space after `,`.

Reply:
Please let me not put my name on the git blame of at least this line :-)

Reply:
Ohh, sorry, thought it was part of the diff.